packages/file/lib/src/backends/memory/memory_file.dart

```diff
@@ -416,19 +416,18 @@ class _FileSink implements io.IOSink {
   Future<void> addStream(Stream<List<int>> stream) {
     _checkNotStreaming();
     _streamCompleter = Completer<void>();
-    void finish() {
-      _streamCompleter!.complete();
-      _streamCompleter = null;
-    }
 
     stream.listen(
       (List<int> data) => _addData(data),
       cancelOnError: true,
       onError: (Object error, StackTrace stackTrace) {
-        _completer.completeError(error, stackTrace);
-        finish();
+        _streamCompleter!.completeError(error, stackTrace);
+        _streamCompleter = null;
+      },
+      onDone: () {
+        _streamCompleter!.complete();
```

Comment:
you might want to check for `isCompleted` here and only call `complete` in that case.

Reply:
It looks like that's unnecessary since `cancelOnError` is true. I can add a test that exercises this if you think it's a good idea.

Here's the test that I used to check this:

```dart
  test(
      'addStream completes on error and does not misbehave if the stream '
      'produces multiple errors and then closes', () async {
    final file = MemoryFileSystem.test().file('foo').openWrite();
    final controller = StreamController<List<int>>()
      ..addError('bar')
      ..addError('baz')
      ..close();
    await expectLater(file.addStream(controller.stream), throwsA('bar'));
    await file.close();
  });
```

Reply:
We could also await the `StreamController.close` for more confidence:

```dart
  test(
      'addStream completes on error and does not misbehave if the stream '
      'produces multiple errors and then closes', () async {
    final file = MemoryFileSystem.test().file('foo').openWrite();
    final controller = StreamController<List<int>>()
      ..addError('bar')
      ..addError('baz');
    final close = controller.close();
    await expectLater(file.addStream(controller.stream), throwsA('bar'));
    await file.close();
    await close;
  });
```

Reply:
Oh, derp, also, since we set `_streamCompleter` to null anywhere we complete it, we'd be unable to complete it multiple times regardless.

Reply:
SGTM, although see my other comment that I do think we should change to using `?.`.